### PR TITLE
[Train] Strip "module." from state dict

### PR DESCRIPTION
**Pull request by Yard1**
Signed-off-by: Antoni Baum <antoni.baum@protonmail.com>

## Why are these changes needed?

This PR adds logic to automatically strip the `"module."` prefix from a user-saved state dict in `TorchCheckpoint`, which is present if a user obtains the state dict from a `DistributedDataParallel` module directly. We already obtain the underlying module if a user saves the model object, so this merely makes the logic consistent.

This PR also edits our examples to remove instances where this operation was conducted in the example itself. This led to issues if `train.torch.prepare_model` was used with `num_workers=1` (eg. on Google Colab), as the module was not wrapped around, thus leading to the `.module` attribute being missing.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/tests/test_torch_trainer.py [hunk 1]
```diff
@@ -111,6 +119,8 @@ def train_func():
     assert predictions.count() == 3
 
 
+# We can't really test for prepare_model here as we can't detect what the user
+# has saved without loading (and thus triggering the exception anyway)
```
Comment (amogkam) [inline python/ray/train/tests/test_torch_trainer.py:123]: for my understanding, can you elaborate on why prepare_model causes this test to fail?

File: python/ray/air/_internal/torch_utils.py [hunk 2]
```diff
@@ -228,3 +228,131 @@ def contains_tensor(obj):
             if contains_tensor(v):
                 return True
     return False
+
+
+# From PyTorch:
```
Comment (amogkam) [inline python/ray/air/_internal/torch_utils.py:233]: can we remove this copyright? the new function is pretty different from the original.

File: python/ray/air/_internal/torch_utils.py [hunk 3]
```diff
@@ -228,3 +228,131 @@ def contains_tensor(obj):
...
+) -> Dict[str, Any]:
+    """Strip the prefix in state_dict, if any and return a new dict.
+
+    Adapted from https://github.com/pytorch/pytorch/blob/\
+c18da597e0bb1c1aecc97c77a73fed1849057fa4/torch/nn/modules/utils.py
+    The original method modified the dict in-place.
+
+    ..note::
+        Given a `state_dict` from a DP/DDP model, a local model can load it by applying
+        `consume_prefix_in_state_dict_if_present(state_dict, "module.")` before calling
```
Comment (amogkam) [inline python/ray/air/_internal/torch_utils.py:327]: this note is no longer relevant?

File: python/ray/air/_internal/torch_utils.py [hunk 4]
```diff
@@ -228,3 +228,131 @@ def contains_tensor(obj):
...
+        :meth:`torch.nn.Module.load_state_dict`.
+
+    Args:
+        state_dict: a state-dict to be loaded to the model.
+        prefix: prefix.
+
+    """
+    copied = False
+
+    keys = sorted(state_dict.keys())
```
Comment (amogkam) [inline python/ray/air/_internal/torch_utils.py:337]: is sorting necessary?

File: python/ray/air/_internal/torch_utils.py [hunk 5]
```diff
@@ -228,3 +228,131 @@ def contains_tensor(obj):
...
+    for key in keys:
+        if key.startswith(prefix):
+            newkey = key[len(prefix) :]
+            if not copied:
+                # We are doing shallow copies here, so the performance
+                # impact should be negligible anyway, but this is
+                # a simple optimization.
+                state_dict = state_dict.copy()
```
Comment (amogkam) [inline python/ray/air/_internal/torch_utils.py:345]: do we need to copy for every single key? we can just do one copy at the end right?

File: python/ray/train/tests/test_torch_trainer.py (hunk 1 above)
Comment (Yard1) [reply python/ray/train/tests/test_torch_trainer.py:123]: `prepare_model` will wrap the model in DDP. If the user doesn't manually unwrap it before saving, DDP will throw an exception after being loaded.

File: python/ray/air/_internal/torch_utils.py (hunk 4 above)
Comment (Yard1) [reply python/ray/air/_internal/torch_utils.py:337]: That's how it was in the original function

File: python/ray/train/tests/test_torch_trainer.py (hunk 1 above)
Comment (amogkam) [reply python/ray/train/tests/test_torch_trainer.py:123]: sorry I guess I mean more- why is it not going through the _encode_dict path?

Comment (Yard1) [reply python/ray/train/tests/test_torch_trainer.py:123]: If a checkpoint is created from directory, we aren't really able to detect what's actually in the files without deserializing them in the first place (which would not only add overhead but also cause the error anyway).

File: python/ray/air/_internal/torch_utils.py (hunk 4 above)
Comment (Yard1) [reply python/ray/air/_internal/torch_utils.py:251]: I'd imagine it's in pytorch for a reason!

File: python/ray/train/tests/test_torch_trainer.py (hunk 1 above)
Comment (amogkam) [reply python/ray/train/tests/test_torch_trainer.py:123]: then for this dir checkpoint, why does it get deserialized in the first place?

Comment (Yard1) [reply python/ray/train/tests/test_torch_trainer.py:123]: We can add prepare_model here but we'd have to unwrap the model before saving anyway, meaning we wouldn't really test anything extra here.

**Review by amogkam — APPROVED** [review]

**Review by richardliaw — APPROVED** [review]

**Pull request merged**